Minor polishing for tplink

rytilahti opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR collects some minor improvements for the tplink based on earlier comments and testing during beta.
* Enable basic effect support on all supported devices
* Set color mode correctly when an effect is active
* Allow current_consumption on light bulbs
* Cleanup some unnecessary code spotted in previous code review rounds

TBD:
* [x] Fix light bulbs power consumption (currently hidden as it's a primary feature on a device with a specialized platform)
* [x] Bump to python-kasa 0.7.0.2: #120940


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

rytilahti committed:
Allow effects on all supported devices

rytilahti committed:
Set color mode correctly when an effect is active

rytilahti committed:
Cleanups based on previous feedback

rytilahti commented on homeassistant/components/tplink/sensor.py:
```diff
@@ -144,21 +143,6 @@ class TPLinkSensorEntity(CoordinatedTPLinkFeatureEntity, SensorEntity):
 
     entity_description: TPLinkSensorEntityDescription
 
-    def __init__(
```
No need for `__init__`s for `sensor` and `switch` as parent's `async_added_to_hass` takes care of updating the state.

rytilahti commented on homeassistant/components/tplink/entity.py:
```diff
@@ -61,8 +61,6 @@
     # update
     "current_firmware_version",
     "available_firmware_version",
-    # fan
-    "fan_speed_level",
```
We have `fan` platform so this is not necessary anymore.

rytilahti committed:
Add allow list for features that can be included even on devices with specialized platforms

rytilahti committed:
Fix tests when an effect is active

sdb9696 commented on homeassistant/components/tplink/entity.py:
```diff
@@ -56,15 +56,21 @@
     DeviceType.Thermostat,
 }
 
+# Features to be always included even when the device type has its own platform
```
```suggestion
# Primary features to be always included even when the device type has its own platform
```

rytilahti replied:
Reworded.

sdb9696 commented on homeassistant/components/tplink/light.py:
```diff
@@ -371,6 +370,7 @@ def _async_update_attrs(self) -> None:
         effect_module = self._effect_module
         if effect_module.effect != LightEffect.LIGHT_EFFECTS_OFF:
             self._attr_effect = effect_module.effect
+            self._attr_color_mode = ColorMode.BRIGHTNESS
```
Do we need to set this back to the correct mode once the effect is switched off? It's slightly confusing because I see there's `_determine_color_mode` in `super()._async_update_attrs` but it's not clear whether that does as much as the `TPLinkLightEntity` constructor.  Is there duplicated logic in there?

rytilahti replied:
I don't think we need to do that as it's taken care of through the `_determine_color_mode`, we just override that for this special case.

I don't see duplicate logic either, as the ctor only initializes the available modes and leaves setting the active one to the `_async_update_attrs` of the corresponding class.

rytilahti committed:
Merge remote-tracking branch 'upstream/dev' into tplink/janitor/cleanup

rytilahti committed:
Remove unnecessary _async_call_update_attrs and fix a comment

rytilahti commented on homeassistant/components/tplink/light.py:
```diff
@@ -215,7 +214,6 @@ def __init__(
         if len(self._attr_supported_color_modes) == 1:
             # If the light supports only a single color mode, set it now
             self._fixed_color_mode = next(iter(self._attr_supported_color_modes))
-        self._async_call_update_attrs()
```
Unnecessary as called in `async_added_to_hass`.

rytilahti replied:
This is now called by the parent's `__init__`, see [`6c971d1` (#120868)](https://github.com/home-assistant/core/pull/120868/commits/6c971d1c7c88889c35ed57e4afe7264261cbd5ae)

rytilahti committed:
Fix voltage unit for feature fixtures, update sensor snapshot

rytilahti committed:
Fix test_update_attrs_fails_in_init test

rytilahti commented on tests/components/tplink/fixtures/features.json:
```diff
@@ -73,7 +73,7 @@
     "type": "Sensor",
     "category": "Primary",
-    "unit": "v",
+    "unit": "V",
```
This caused a warning due to incorrect unit for the device class.

rytilahti commented on tests/components/tplink/snapshots/test_sensor.ambr:
```diff
@@ -146,7 +146,7 @@
     'supported_features': 0,
     'translation_key': 'battery_level',
     'unique_id': '123456789ABCDEFGH_battery_level',
-    'unit_of_measurement': '%',
```
We miss the units & suggested display precisions as their values are first initialized after `async_add_to_hass` has been called & this is not currently done.

rytilahti committed:
Remove async_added_to_hass and use _async_call_update_attrs already late in __init__

This is necessary as setting the attributes in async_added_to_hass is too late for snapshot testing

rytilahti committed:
cleanup
